iModelCore/ECPresentation/Source/Content/ContentSpecificationsHandler.cpp

```diff
@@ -226,22 +256,98 @@ static ECSchemaHelper::RelationshipPathsResponse FindRelatedPropertyPaths(Conten
-    auto flatSpecs = FlattenedRelatedPropertiesSpecification::Create(params.GetRelatedPropertySpecs(), RelatedPropertiesSpecificationScopeInfo(params.GetScopeCategorySpecifications()));
-    for (ContentModifierCP modifier : GetContext().GetRulesPreprocessor().GetContentModifiers())
+    bvector<std::unique_ptr<FlattenedRelatedPropertiesSpecification>> specs;
+    for (ContentModifierCP modifier : modifiers)
         {
-        if (params.GetSourceClassInfo().GetSelectClass().GetClass().Is(modifier->GetSchemaName().c_str(), modifier->GetClassName().c_str()))
+        ECClassCP modifierClass = helper.GetECClass(modifier->GetSchemaName().c_str(), modifier->GetClassName().c_str());
+        if (modifierClass == nullptr)
+            {
+            DIAGNOSTICS_LOG(DiagnosticsCategory::Content, LOG_INFO, LOG_ERROR, Utf8PrintfString("Content modifier specifies non-existing class: %s.%s.",
+               modifier->GetSchemaName().c_str(), modifier->GetClassName().c_str()));
+            continue;
+            }
+        bool propertyClassMatchesModifierClass = selectClass.GetClass().Is(modifierClass);
+        bool modifierClassMatchesPropertyClass = selectClass.IsSelectPolymorphic() && modifierClass->Is(&selectClass.GetClass());
+        if (propertyClassMatchesModifierClass || modifierClassMatchesPropertyClass)
             {
             DiagnosticsHelpers::ReportRule(*modifier);
-            ContainerHelpers::MovePush(flatSpecs, FlattenedRelatedPropertiesSpecification::Create(modifier->GetRelatedProperties(), RelatedPropertiesSpecificationScopeInfo(modifier->GetPropertyCategories())));
+            ContainerHelpers::MovePush(specs, FlattenedRelatedPropertiesSpecification::Create(modifier->GetRelatedProperties(), RelatedPropertiesSpecificationScopeInfo(modifier->GetPropertyCategories())));
+            }
+        }
+    return specs;
+    }
+
+/*---------------------------------------------------------------------------------**//**
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+static bvector<std::unique_ptr<FlattenedRelatedPropertiesSpecification>> CreateForFlatSpecFromModifiers(RelatedPropertiesSpecificationCR spec, bvector<ContentModifierCP> modifiers, ECSchemaHelper const& helper)
+    {
+    bvector<std::unique_ptr<FlattenedRelatedPropertiesSpecification>> specs;
+    if (!spec.AllPropertiesIncluded())
+        return specs;
+
+    int handledModifiersInPrevIteration = -1;
+    while (handledModifiersInPrevIteration != 0)
+        {
+        handledModifiersInPrevIteration = 0;
+        for (ContentModifierCP& modifier : modifiers)
+            {
+            ECClassCP modifierClass = nullptr;
+            if (modifier == nullptr || !modifier->ShouldApplyOnNestedContent() || nullptr == (modifierClass = helper.GetECClass(modifier->GetSchemaName().c_str(), modifier->GetClassName().c_str())))
+                continue;

```

Comment:
I'd change this to the following:

```cpp
for (auto& modifier : modifiers)
    {
    if (!modifier || !modifier->ShouldApplyOnNestedContent())
        continue;

    ECClassCP modifierClass = helper.GetECClass(modifier->GetSchemaName().c_str(), modifier->GetClassName().c_str());
    if (modifierClass == nullptr)
        {
        DIAGNOSTICS_LOG(DiagnosticsCategory::Content, LOG_INFO, LOG_ERROR, Utf8PrintfString("Content modifier %s specifies non-existing class: %s.%s.",
            modifier->GetSchemaName().c_str(), modifier->GetClassName().c_str()));
        continue;
        }
```

Reply:
Fixed. Just in your example  `Utf8PrintfString("Content modifier %s specifies non-existing class: %s.%s.", modifier->GetSchemaName().c_str(), modifier->GetClassName().c_str())` what is the first `%s`? As I understand, these other `%s`s are `schemaName.className`.

Reply:
Good catch - I copied it from `ContentClassesLocater` and it had this issue there.